src/DynamicRupture/Output/Base.cpp

```diff
@@ -290,6 +425,170 @@ void Base::calcFaultOutput(const OutputType type, OutputData& outputData, double
+      local.faceAlignedValuesPlus[6] + (local.p - local.faceAlignedValuesPlus[0]) * uVelDivisor;
+
+  real missingSigmaValues = (local.p - local.faceAlignedValuesPlus[0]);
+  missingSigmaValues *= (1.0 - 2.0 * std::pow(local.waveSpeedsPlus->sWaveVelocity /
+                                                  local.waveSpeedsPlus->pWaveVelocity,
+                                              2));
+
+  local.yyStress = local.faceAlignedValuesPlus[1] + missingSigmaValues;
+  local.zzStress = local.faceAlignedValuesPlus[2] + missingSigmaValues;
+  local.yzStress = local.faceAlignedValuesPlus[4];
+
+  local.tracEla =
+      std::sqrt(std::pow(local.sXY + local.xyStress, 2) + std::pow(local.sXZ + local.xzStress, 2));
+}
+
+void Base::computeLocalTraction(real strength) {
+  if (local.tracEla > std::abs(strength)) {
+    local.xyTraction = ((local.sXY + local.xyStress) / local.tracEla) * strength;
+    local.xzTraction = ((local.sXZ + local.xzStress) / local.tracEla) * strength;
+
+    // update stress change
+    local.xyTraction -= local.sXY;
+    local.xzTraction -= local.sXZ;
+  } else {
+    local.xyTraction = local.xyStress;
+    local.xzTraction = local.xzStress;
+  }
+}
+
+void Base::computeSlipAndRate(std::array<real, 6>& rotatedTraction,
+                              std::array<real, 6>& rotatedLocalStress) {
+  real sFactorPlus = local.waveSpeedsPlus->sWaveVelocity * local.waveSpeedsPlus->density;
+  real sFactorMinus = local.waveSpeedsMinus->sWaveVelocity * local.waveSpeedsMinus->density;
+
+  local.srS =
+      -(1.0 / sFactorPlus + 1.0 / sFactorMinus) * (rotatedTraction[3] - rotatedLocalStress[3]);
+  local.srD =
+      -(1.0 / sFactorPlus + 1.0 / sFactorMinus) * (rotatedTraction[5] - rotatedLocalStress[5]);
+}
+
+void Base::computeSlipAndRate(const double* tangent1,
+                              const double* tangent2,
+                              const double* strike,
+                              const double* dip) {
+  local.srS = static_cast<real>(0.0);
+  local.srD = static_cast<real>(0.0);
+
+  for (size_t i = 0; i < 3; ++i) {
+    real factorMinus = (local.faceAlignedValuesMinus[7] * tangent1[i] +
+                        local.faceAlignedValuesMinus[8] * tangent2[i]);
+
+    real factorPlus = (local.faceAlignedValuesPlus[7] * tangent1[i] +
+                       local.faceAlignedValuesPlus[8] * tangent2[i]);
+
+    local.srS += factorMinus * strike[i] - factorPlus * strike[i];
+    local.srD += factorMinus * dip[i] - factorPlus * dip[i];
+  }
+}
+
+int Base::getClosestInternalGp(int nearestGpIndex, int nPoly) {
+  int i1 = int((nearestGpIndex - 1) / (nPoly + 2)) + 1;
+  int j1 = (nearestGpIndex - 1) % (nPoly + 2) + 1;
+  if (i1 == 1) {
+    i1 = i1 + 1;
+  } else if (i1 == (nPoly + 2)) {
+    i1 = i1 - 1;
+  }
+
+  if (j1 == 1) {
+    j1 = j1 + 1;
+  } else if (j1 == (nPoly + 2)) {
+    j1 = j1 - 1;
+  }
+  return (i1 - 1) * (nPoly + 2) + j1;
+}
+
+real Base::computeRuptureVelocity(Eigen::Matrix<real, 2, 2>& jacobiT2d) {
```

Comment:
What does `T2d` stand for?

Reply:
It is something that I wanted to ask you. Here is the corresponding snippet from the Fortran code.
```
                grad2d(1,1) = dot_product(NormalVect_s(:),xab)
                grad2d(1,2) = dot_product(NormalVect_t(:),xab)
                grad2d(2,1) = dot_product(NormalVect_s(:),xac)
                grad2d(2,2) = dot_product(NormalVect_t(:),xac)
                CALL MatrixInverse2x2(JacobiT2d,grad2d)

                dt_dx1 = JacobiT2d(1,1) * dt_dchi + JacobiT2d(1,2) * dt_dtau
                dt_dy1 = JacobiT2d(2,1) * dt_dchi + JacobiT2d(2,2) * dt_dtau
                Slowness = sqrt(dt_dx1**2 + dt_dy1**2)
```

Reply:
so JacobiT2d(1,1) is dChi/dx1 
(1.2) is dTau/dx1, etc.
JacobiT2d(2,1) is dChi/dx2
and (2,2) dTau/dx2.